src/shogun/io/openml/OpenMLRun.cpp

```diff
@@ -0,0 +1,183 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Gil Hoben
+ */
+
+#include <shogun/evaluation/CrossValidationStorage.h>
+#include <shogun/io/openml/OpenMLRun.h>
+#include <shogun/io/openml/ShogunOpenML.h>
+#include <shogun/labels/Labels.h>
+#include <shogun/machine/Machine.h>
+
+using namespace shogun;
+
+std::shared_ptr<OpenMLRun> OpenMLRun::run_model_on_task(
+    std::shared_ptr<CSGObject> model, std::shared_ptr<OpenMLTask> task)
+{
+	SG_SNOTIMPLEMENTED
+	return std::shared_ptr<OpenMLRun>();
+}
+
+std::shared_ptr<OpenMLRun> OpenMLRun::run_flow_on_task(
+    std::shared_ptr<OpenMLFlow> flow, std::shared_ptr<OpenMLTask> task)
+{
+	auto data = task->get_dataset();
+	std::shared_ptr<CFeatures> features = nullptr;
+	std::shared_ptr<CLabels> labels = nullptr;
+
+	auto model = ShogunOpenML::flow_to_model(std::move(flow), true);
+
+	labels = data->get_labels();
+	features = data->get_features(data->get_default_target_attribute());
+
+	auto storage = std::make_shared<CrossValidationStorage>();
+
+	if (task->get_split()->contains_splits())
+	{
+		auto machine = std::dynamic_pointer_cast<CMachine>(model);
+		if (!machine)
+		{
+			SG_SERROR("INTERNAL ERROR: failed to cast model to machine!\n")
+		}
+		auto train_idx = task->get_train_indices();
+		auto test_idx = task->get_test_indices();
+
+		auto xval_storage = std::make_shared<CrossValidationStorage>();
+		xval_storage->set_num_folds(task->get_num_fold());
+		xval_storage->set_num_runs(task->get_num_repeats());
+		machine->set_store_model_features(true);
+
+		// copied/adapted from crossvalidation
+		for (auto repeat_idx : range(task->get_num_repeats()))
+		{
+			for (auto fold_idx : range(task->get_num_fold()))
+			{
+				auto* fold = new CrossValidationFoldStorage();
+				SG_REF(fold)
+
+				auto cloned_machine = (CMachine*)machine->clone();
+
+				// TODO while these are not used through const interfaces,
+				//  we unfortunately have to clone, even though these could be
+				//  shared
+				auto features_clone = (CFeatures*)features->clone();
```

Comment:
I see the TODO is taken from the other xval. But we shouldn't merge this.
What about you create a new features instance with the same data instead of cloning it. Same with the labels

Reply:
yes, I did some shameless copy/paste here :D

Reply:
Actually that makes me think. Can't we re-use the existing code? Could refactor it to allow that. Because having it twice is meh

Reply:
sure thing, I mean what I wrote is the boilerplate, so can just add a utility function in crossvalidation that does a fold train/test/eval and takes subset as an argument

Reply:
for example, yes :)
Maybe we can do that in Berlin ...